File: moqt/internal/message/session_update.go
package message

import (
	"io"
)

type SessionUpdateMessage struct {
	/*
	 * Versions selected by the server
	 */
	Bitrate uint64
}

func (sum SessionUpdateMessage) Len() int {
	return VarintLen(sum.Bitrate)
}

func (sum SessionUpdateMessage) Encode(w io.Writer) error {
	msgLen := sum.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, sum.Bitrate)

	_, err := w.Write(b)
	return err
}

func (sum *SessionUpdateMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	sum.Bitrate = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: interop/web/src/message/session_update.ts
import { Writer, Reader } from "../io";
import { varintLen } from "../io/len";

export class SessionUpdateMessage {
    bitrate: bigint;

    constructor(bitrate: bigint) {
        this.bitrate = bitrate;
    }

    length(): number {
        return varintLen(this.bitrate);
    }

    static async encode(writer: Writer, bitrate: bigint): Promise<[SessionUpdateMessage?, Error?]> {
        const msg = new SessionUpdateMessage(bitrate);
        writer.writeVarint(BigInt(msg.length()));
        writer.writeVarint(bitrate);
        const err = await writer.flush();
        if (err) {
            return [undefined, err];
        }
        return [msg, undefined];
    }

    static async decode(reader: Reader): Promise<[SessionUpdateMessage?, Error?]> {
        const [len, err] = await reader.readVarint();
        if (err) {
            return [undefined, new Error("Failed to read length for SessionUpdateMessage")];
        }
        const [varint, err2] = await reader.readVarint();
        if (err2) {
            return [undefined, new Error("Failed to read bitrate for SessionUpdateMessage: " + err2.message)];
        }

        return [new SessionUpdateMessage(varint), undefined];
    }
}